### PR TITLE
[GEOT-7406] Vector Mosaic Cache granule connection in memory to avoid  redundant initialization and improve granule iterator cleanup

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOT-7406](https://osgeo-org.atlassian.net/browse/GEOT-7406) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java [hunk 1]
```diff
@@ -361,14 +366,17 @@ private Object toConnectionParameter(String connectionParameterKey, String param
...
      */
-    private void findDataStore(VectorMosaicGranule granule, boolean isSampleForType) {
+    private Optional<DataStore> findDataStore(
+            VectorMosaicGranule granule, boolean isSampleForType) {
         VectorMosaicGranuleStoreFinder finder = null;
+        Optional<DataStore> dataStore = Optional.empty();
         if (granuleTracker == null) {
             finder = new VectorMosaicGranuleStoreFinderImpl(preferredSPI);
-            finder.findDataStore(granule, isSampleForType);
+            dataStore = finder.findDataStore(granule, isSampleForType);
```
Comment (aaime) [inline modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java:375]: Repeated line, could be out of the if.

File: modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicReaderTest.java [hunk 2]
```diff
@@ -264,6 +283,9 @@ public void testLikeFilter() throws Exception {
         assertNotEquals(
                 Filter.INCLUDE,
                 ((VectorMosaicFeatureSource) featureSource).filterTracker.getGranuleFilter());
-        assertEquals(3, tracker.size()); // all indexes are hit, so all granules are checked
+        assertEquals(
+                3,
+                tracker.getGranuleNames()
+                        .size()); // all indexes are hit, so all granules are checked
```
Comment (aaime) [inline modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicReaderTest.java:289]: Play nice with the formatter. I've commented on this one, but there are other cases around like this.

```suggestion
        // all indexes are hit, so all granules are checked
        assertEquals(3,  tracker.getGranuleNames().size()); 
```

**Review by aaime — CHANGES_REQUESTED** [review]
Generally speaking, I'd like to see the code changed so that testing support is all inside testing itself, without the main code having to do ifs to handle it. Or at least centralized  more? Like, the store finder implemnetation could be a package private field, and testing would be able to replace it with the monitoring implemetation. Also wondering if the monitoring could be just a wrapper, so that you don't have to duplicate as much code.

File: modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicReaderTest.java (hunk 2 above)
Comment (turingtestfail) [reply modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicReaderTest.java]: fixed

File: modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java (hunk 1 above)
Comment (turingtestfail) [reply modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java]: Fixed

File: modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java [hunk 3]
```diff
@@ -270,11 +272,18 @@ protected boolean isNotMandatoryIndexType(AttributeDescriptor descriptor) {
...
+        DataStore dataStore = null;
         validateAndLoadConnectionStringProperties(granule);
-        findDataStore(granule, isSampleForType);
-        populateGranuleTypeName(granule);
+        Optional<DataStore> dataStoreOptional = findDataStore(granule, isSampleForType);
+        if (dataStoreOptional.isPresent()) {
+            dataStore = dataStoreOptional.get();
+        } else {
+            throw new IOException("No data store found for granule");
+        }
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java]: fixed

File: modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicGranuleStoreFinderMonitoring.java [hunk 4]
```diff
@@ -71,5 +76,35 @@ public void findDataStore(VectorMosaicGranule granule, boolean isSampleForSchema
...
         }
+        return Optional.ofNullable(dataStore);
+    }
+
+    public static class GranuleTracker {
+        private Set<String> granuleNames = new HashSet<>();
+        private int count = 0;
+
+        public void incrementAccessCount() {
+            count++;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicGranuleStoreFinderMonitoring.java]: Fixed

File: modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java [hunk 5]
```diff
@@ -57,7 +57,7 @@ public class VectorMosaicFeatureSource extends ContentFeatureSource {
     private final String preferredSPI;
     private final String connectionParameterKey;
     protected FilterTracker filterTracker = new FilterTracker();
-    protected Set<String> granuleTracker;
+    protected VectorMosaicGranuleStoreFinderMonitoring.GranuleTracker granuleTracker;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java]: fixed

**Comment by turingtestfail** [comment]
> Generally speaking, I'd like to see the code changed so that testing support is all inside testing itself, without the main code having to do ifs to handle it. Or at least centralized more? Like, the store finder implemnetation could be a package private field, and testing would be able to replace it with the monitoring implemetation. Also wondering if the monitoring could be just a wrapper, so that you don't have to duplicate as much code.

Ended up going with the tracker as package private and with removing the monitoring finder implementation altogether in order to minimize code duplication.

**Review by aaime — APPROVED** [review]

**Pull request merged**